### PR TITLE
[BLOCK-1459] 1st hand purchase actions

### PR DESCRIPTION
**Pull request by justefg**

### CODE REVIEW

File: docs/contracts/NFT Contract/NFT Actions/delprchsreq.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+TBA
+
+## Action Parameters
+
+**Action Interface**
+
+| Property Name            | C++ Type   | JavaScript Type |
+| ------------------------ | -----------| --------------- |
+| token_factory_id         | uint64_t   | number          |
+| index                    | uint64_t   | number          |
```
Comment (igor-sikachyna) [inline docs/contracts/NFT Contract/NFT Actions/delprchsreq.md:22]: Don't forget to add the descriptions for the action fields

File: docs/contracts/NFT Contract/NFT Actions/purchase.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+| Property Name       | C++ Type                       | JavaScript Type       |
+| ------------------- | ------------------------------ | --------------------- |
+| token_factory_id    | uint64_t                       | number                |
+| index               | uint64_t                       | number                |
+| max_price           | asset                          | string                |
+| buyer               | optional name                  | string / null         |
+| receiver            | optional asset                 | string / null         |
+| promoter_id         | optional name                  | string / null         |
+| user_uniqs          | optional provided_user_uniqs   | Array / null          |
```
Comment (igor-sikachyna) [inline docs/contracts/NFT Contract/NFT Actions/purchase.md:27]: For optional types you could still put angled braces if you put the \ in front of them like optional\<name>

File: docs/contracts/NFT Contract/NFT Actions/purchase.a.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+In some cases a token factory may require certain uniqs to exist in the user inventory table to enable the user to purchase a uniq.
+
+Think of it like a pre-requisite or an entry ticket to purchasing other uniqs.
+
+When a uniq is being purchased it goes through our `verify_user_uniqs` function that looks into the buyer's inventory and verifies that the token factories required uniqs matches the user supplied uniqs. The function checks that the `user supplied uniqs` from the user matches the `factory required uniqs`. The function checks that the user is passing uniqs that have the correct token factory id, and checks that the user **is not** passing irrelavant uniqs.
+
+It also ensures that the strategy that is being passed for each uniq matches the strategy used by the factory for the specific token factory id.
+
+Strategy meaning: `0` just check, `1` burn the uniq and `2` transferring the uniq out of the user inventory.
```
Comment (Adam-Ultra) [inline docs/contracts/NFT Contract/NFT Actions/purchase.a.md:45]: `0` just check => `0` just check ownership of the provided nft

File: docs/contracts/NFT Contract/NFT Actions/purchase.a.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+| receiver         | eosio::asset                        | string          | Account that will receive the Uniq from this purchase                                                                                                                               |
+| promoter_id      | std::optional\<eosio::name>         | string / null   | Optional promoter of the purchase transaction. If no promoter is provided then the default promoter specified in `saleshrlmcfg` (scope `0`) will be used if present                 |
+| user_uniqs       | std::optional\<provided_user_uniqs> | Array / null    | List of uniqs the buyer is willing to provide for this purchase option to either be taken from him or to just verify their presence. Refer to `provided_user_uniqs` breakdown below |
+| memo             | std::string                         | string          | A short operation description                                                                                                                                                       |
+
+**provided_user_uniqs Interface**
```
Comment (Adam-Ultra) [inline docs/contracts/NFT Contract/NFT Actions/purchase.a.md:79]: provided_user_uniqs Interface is a vector of uniqs_count_internal, which is the following sturcuture

File: docs/contracts/NFT Contract/nft-tables.md
[the remark this answers is not in the document]
Comment (igor-sikachyna) [reply docs/contracts/NFT Contract/nft-tables.md:1]: The sidebar does not add stuff that is wrapped in experimental tag

File: docs/contracts/NFT Contract/overview.md [hunk 5]
```diff
@@ -134,3 +134,16 @@ The **setconrecv** action allows the _Asset Manager_ to set the token factory co
...
+The _Asset Manager_ can add the factory first-hand purchase requirements
+
+The **setprchsreq.a** action allows the _Asset Manager_ to set a purchase requirement for the factory which can then be used by any user to purchase a token from the factory directly using **purchase.a** action.
+
+Multiple purchase requirements can be specified for a single factory. In addition to a simple fungible token price (UOS or USD) asset manager can specify the price for an individual purchase option using uniqs from other factories. Those uniqs can be either burnt or transferred to a specified account or simply verify their presence.
+
+-   [setprchsreq.a - set purchase requirement](./NFT%20Actions/setprchsreq.a.md)
+-   [delprchsreq.a - delete purchase requirement](./NFT%20Actions/delprchsreq.a.md)
+-   [purchase.a - purchase a token](./NFT%20Actions/purchase.a.md)
+-   [fctrprchs.a](./nft-tables.md#fctrprchs-a)
```
[the remark this answers is not in the document]
Comment (igor-sikachyna) [reply docs/contracts/NFT Contract/overview.md:149]: Will fix the link. Other table links in this document are also missing the brief. I can add it for this one though

[the remark this answers is not in the document]
Comment (igor-sikachyna) [reply docs/contracts/NFT Contract/overview.md:149]: I just checked and the link seems to be working fine both in VS code and when running documentation. Are you sure you checked the latest version?

**Review by Adam-Ultra — APPROVED** [review]

**Review by igor-sikachyna — APPROVED** [review]

**Pull request merged**